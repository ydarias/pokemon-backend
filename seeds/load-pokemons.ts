import { DataSource, EntityManager } from 'typeorm';
import {
  AttackEntity,
  EvolutionEntity,
  EvolutionRequirementsEntity,
  MeasurementEntity,
  PokemonEntity,
  TypeEntity,
} from '../src/modules/pokemon-catalog/infra/pokemon.entity';

import * as pokemons from './pokemons.json';

async function persistTypesInCollection(types: string[], entityManager: EntityManager): Promise<TypeEntity[]> {
  const entities: TypeEntity[] = [];
  for (let i = 0; i < types.length; i++) {
    const type = new TypeEntity();
    type.name = types[i];

    entities.push(await entityManager.save(type));
  }
  return entities;
}

async function persistAttacksInCollection(attacks: any, entityManager: EntityManager): Promise<AttackEntity[]> {
  const entities: AttackEntity[] = [];
  for (let i = 0; i < attacks.length; i++) {
    const attack = new AttackEntity();
    attack.type = attacks[i].type;
    attack.name = attacks[i].name;
    attack.damage = attacks[i].damage;

    entities.push(await entityManager.save(attack));
  }
  return entities;
}

async function persistEvolutionsInCollection(
  evolutions: any,
  entityManager: EntityManager,
): Promise<EvolutionEntity[]> {
  if (evolutions) {
    const entities: EvolutionEntity[] = [];
    for (let i = 0; i < evolutions.length; i++) {
      const evolution = new EvolutionEntity();
      evolution.id = evolutions[i].id.toString().padStart(3, '0');
      evolution.name = evolutions[i].name;

      entities.push(await entityManager.save(evolution));
    }
    return entities;
  }

  return [];
}

function parseWeightEntity(p) {
  const weightEntity = new MeasurementEntity();
  weightEntity.maximum = Number.parseFloat(p.weight.maximum.replace('kg', ''));
  weightEntity.minimum = Number.parseFloat(p.weight.minimum.replace('kg', ''));
  weightEntity.unit = 'kg';

  return weightEntity;
}

function parseHeightEntity(p) {
  const heightEntity = new MeasurementEntity();
  heightEntity.maximum = Number.parseFloat(p.height.maximum.replace('m', ''));
  heightEntity.minimum = Number.parseFloat(p.height.minimum.replace('m', ''));
  heightEntity.unit = 'm';

  return heightEntity;
}

function parseEvolutionRequirementsEntity(p) {
  const evolutionRequirementsEntity = new EvolutionRequirementsEntity();
  evolutionRequirementsEntity.name = p.evolutionRequirements.name;
  evolutionRequirementsEntity.amount = p.evolutionRequirements.amount;

  return evolutionRequirementsEntity;
}

export const loadPokemonsData = async (datasource: DataSource) => {
  const entityManager = datasource.manager;

  for (let i = 0; i < pokemons.length; i++) {
    const p = pokemons[i];

    await persistTypesInCollection(p.types, entityManager);

    const pokemonEntity = new PokemonEntity();
    pokemonEntity.id = p.id;
    pokemonEntity.name = p.name;
    pokemonEntity.nameForSearch = p.name.toLowerCase();
    pokemonEntity.classification = p.classification;
    pokemonEntity.types = p.types;
    pokemonEntity.resistant = p.resistant;
    pokemonEntity.weaknesses = p.weaknesses;
    pokemonEntity.fleeRate = p.fleeRate;
    pokemonEntity.maxCP = p.maxCP;
    pokemonEntity.maxHP = p.maxHP;
    pokemonEntity.fastAttacks = await persistAttacksInCollection(p.attacks.fast, entityManager);
    pokemonEntity.specialAttacks = await persistAttacksInCollection(p.attacks.special, entityManager);
    pokemonEntity.evolutions = await persistEvolutionsInCollection(p.evolutions, entityManager);
    pokemonEntity.previousEvolutions = await persistEvolutionsInCollection(p['Previous evolution(s)'], entityManager);
    pokemonEntity.weight = await entityManager.save(parseWeightEntity(p));
    pokemonEntity.height = await entityManager.save(parseHeightEntity(p));

    if (p.evolutionRequirements) {
      pokemonEntity.evolutionRequirements = await entityManager.save(parseEvolutionRequirementsEntity(p));
    }

    await entityManager.save(pokemonEntity);
  }
};
