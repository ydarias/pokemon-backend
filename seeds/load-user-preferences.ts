import { DataSource } from 'typeorm';
import { UserPreferencesEntity } from '../src/modules/user-pokedex/infra/user-preferences.entity';

export const loadUserPreferences = async (datasource: DataSource) => {
  const entityManager = datasource.manager;

  await entityManager.delete(UserPreferencesEntity, {});

  const userPreferencesEntity = new UserPreferencesEntity();
  userPreferencesEntity.userID = 'default-user';
  userPreferencesEntity.favoritePokemons = ['034', '035', '037', '038'];

  await entityManager.save(userPreferencesEntity);
};
