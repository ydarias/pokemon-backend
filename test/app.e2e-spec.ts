import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { AppModule } from '../src/app.module';
import { MockedPokemons } from '../src/utils/tests/pokemons';
import { loadPokemonsData } from '../seeds/load-pokemons';
import { loadUserPreferences } from '../seeds/load-user-preferences';
import { DataSource } from 'typeorm';

describe('AppController (e2e)', () => {
  let app: INestApplication;

  jest.setTimeout(15000);

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();

    const datasource = app.get(DataSource);
    await loadPokemonsData(datasource);
    await loadUserPreferences(datasource);
  });

  it('should support requesting a pokemon by its ID', async () => {
    const response = await request(app.getHttpServer()).get('/pokemons/025').expect(200);

    expect(response.body).toMatchObject(MockedPokemons.pikachuResponse());
  });

  it('should support requesting a pokemon by its name', async () => {
    const response = await request(app.getHttpServer()).get('/pokemons/name/charizard').expect(200);

    expect(response.body).toMatchObject(MockedPokemons.charizardResponse());
  });

  it('should support requesting a page of pokemons (default order is by ID)', async () => {
    const response = await request(app.getHttpServer()).get('/pokemons?limit=3&skip=3').expect(200);

    expect(response.body).toMatchSnapshot();
  });

  it('should support requesting a page of pokemons filtered by type (default order is by ID)', async () => {
    const response = await request(app.getHttpServer()).get('/pokemons?limit=3&type=Fire').expect(200);

    expect(response.body).toMatchSnapshot();
  });

  it('should support adding favorite pokemons to a user preferences', async () => {
    await request(app.getHttpServer())
      .put('/me/favorites')
      .send({
        add: ['001', '002'],
        remove: ['034'],
      })
      .expect(200);

    const response = await request(app.getHttpServer()).get('/me').expect(200);

    expect(response.body).toStrictEqual({
      userID: 'default-user',
      favoritePokemons: ['035', '037', '038', '001', '002'],
    });
  });

  it('should support requesting a page of pokemons filtered by type and favorites', async () => {
    const response = await request(app.getHttpServer()).get('/pokemons?limit=20&type=Fire&favorites=true').expect(200);

    expect(response.body).toMatchSnapshot();
  });

  it('should support requesting the types of pokemons (default order alphabetical)', async () => {
    const response = await request(app.getHttpServer()).get('/pokemons/literals/types').expect(200);

    expect(response.body).toStrictEqual([
      'Bug',
      'Dragon',
      'Electric',
      'Fairy',
      'Fighting',
      'Fire',
      'Flying',
      'Ghost',
      'Grass',
      'Ground',
      'Ice',
      'Normal',
      'Poison',
      'Psychic',
      'Rock',
      'Steel',
      'Water',
    ]);
  });
});
