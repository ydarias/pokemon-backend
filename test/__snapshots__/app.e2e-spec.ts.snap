// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) should support requesting a page of pokemons (default order is by ID) 1`] = `
Object {
  "items": Array [
    Object {
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
          Object {
            "damage": 6,
            "name": "Scratch",
            "type": "Normal",
          },
        ],
        "special": Array [
          Object {
            "damage": 30,
            "name": "Flame Burst",
            "type": "Fire",
          },
          Object {
            "damage": 25,
            "name": "Flame Charge",
            "type": "Fire",
          },
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
        ],
      },
      "classification": "Lizard Pokémon",
      "evolutionRequirements": Object {
        "amount": 25,
        "name": "Charmander candies",
      },
      "evolutions": Array [
        Object {
          "id": "006",
          "name": "Charizard",
        },
        Object {
          "id": "005",
          "name": "Charmeleon",
        },
      ],
      "fleeRate": 0.1,
      "height": Object {
        "maximum": "0.68m",
        "minimum": "0.53m",
      },
      "id": "004",
      "maxCP": 841,
      "maxHP": 955,
      "name": "Charmander",
      "resistant": Array [
        "Fire",
        "Grass",
        "Ice",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
      ],
      "weaknesses": Array [
        "Water",
        "Ground",
        "Rock",
      ],
      "weight": Object {
        "maximum": "9.56kg",
        "minimum": "7.44kg",
      },
    },
    Object {
      "Previous evolution(s)": Array [
        Object {
          "id": "004",
          "name": "Charmander",
        },
      ],
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
          Object {
            "damage": 6,
            "name": "Scratch",
            "type": "Normal",
          },
        ],
        "special": Array [
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
          Object {
            "damage": 30,
            "name": "Flame Burst",
            "type": "Fire",
          },
          Object {
            "damage": 40,
            "name": "Fire Punch",
            "type": "Fire",
          },
        ],
      },
      "classification": "Flame Pokémon",
      "evolutionRequirements": Object {
        "amount": 100,
        "name": "Charmander candies",
      },
      "evolutions": Array [
        Object {
          "id": "006",
          "name": "Charizard",
        },
      ],
      "fleeRate": 0.07,
      "height": Object {
        "maximum": "1.24m",
        "minimum": "0.96m",
      },
      "id": "005",
      "maxCP": 1411,
      "maxHP": 1557,
      "name": "Charmeleon",
      "resistant": Array [
        "Fire",
        "Grass",
        "Ice",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
      ],
      "weaknesses": Array [
        "Water",
        "Ground",
        "Rock",
      ],
      "weight": Object {
        "maximum": "21.38kg",
        "minimum": "16.63kg",
      },
    },
    Object {
      "Previous evolution(s)": Array [
        Object {
          "id": "005",
          "name": "Charmeleon",
        },
        Object {
          "id": "004",
          "name": "Charmander",
        },
      ],
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
          Object {
            "damage": 9,
            "name": "Wing Attack",
            "type": "Flying",
          },
        ],
        "special": Array [
          Object {
            "damage": 35,
            "name": "Dragon Claw",
            "type": "Dragon",
          },
          Object {
            "damage": 100,
            "name": "Fire Blast",
            "type": "Fire",
          },
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
        ],
      },
      "classification": "Flame Pokémon",
      "fleeRate": 0.05,
      "height": Object {
        "maximum": "1.91m",
        "minimum": "1.49m",
      },
      "id": "006",
      "maxCP": 2413,
      "maxHP": 2602,
      "name": "Charizard",
      "resistant": Array [
        "Fire",
        "Grass",
        "Fighting",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
        "Flying",
      ],
      "weaknesses": Array [
        "Water",
        "Electric",
        "Rock",
      ],
      "weight": Object {
        "maximum": "101.81kg",
        "minimum": "79.19kg",
      },
    },
  ],
  "meta": Object {
    "count": 151,
    "limit": 3,
    "skip": 3,
  },
}
`;

exports[`AppController (e2e) should support requesting a page of pokemons filtered by type (default order is by ID) 1`] = `
Object {
  "items": Array [
    Object {
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
          Object {
            "damage": 6,
            "name": "Scratch",
            "type": "Normal",
          },
        ],
        "special": Array [
          Object {
            "damage": 30,
            "name": "Flame Burst",
            "type": "Fire",
          },
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
          Object {
            "damage": 25,
            "name": "Flame Charge",
            "type": "Fire",
          },
        ],
      },
      "classification": "Lizard Pokémon",
      "evolutionRequirements": Object {
        "amount": 25,
        "name": "Charmander candies",
      },
      "evolutions": Array [
        Object {
          "id": "005",
          "name": "Charmeleon",
        },
        Object {
          "id": "006",
          "name": "Charizard",
        },
      ],
      "fleeRate": 0.1,
      "height": Object {
        "maximum": "0.68m",
        "minimum": "0.53m",
      },
      "id": "004",
      "maxCP": 841,
      "maxHP": 955,
      "name": "Charmander",
      "resistant": Array [
        "Fire",
        "Grass",
        "Ice",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
      ],
      "weaknesses": Array [
        "Water",
        "Ground",
        "Rock",
      ],
      "weight": Object {
        "maximum": "9.56kg",
        "minimum": "7.44kg",
      },
    },
    Object {
      "Previous evolution(s)": Array [
        Object {
          "id": "004",
          "name": "Charmander",
        },
      ],
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 6,
            "name": "Scratch",
            "type": "Normal",
          },
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
        ],
        "special": Array [
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
          Object {
            "damage": 30,
            "name": "Flame Burst",
            "type": "Fire",
          },
          Object {
            "damage": 40,
            "name": "Fire Punch",
            "type": "Fire",
          },
        ],
      },
      "classification": "Flame Pokémon",
      "evolutionRequirements": Object {
        "amount": 100,
        "name": "Charmander candies",
      },
      "evolutions": Array [
        Object {
          "id": "006",
          "name": "Charizard",
        },
      ],
      "fleeRate": 0.07,
      "height": Object {
        "maximum": "1.24m",
        "minimum": "0.96m",
      },
      "id": "005",
      "maxCP": 1411,
      "maxHP": 1557,
      "name": "Charmeleon",
      "resistant": Array [
        "Fire",
        "Grass",
        "Ice",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
      ],
      "weaknesses": Array [
        "Water",
        "Ground",
        "Rock",
      ],
      "weight": Object {
        "maximum": "21.38kg",
        "minimum": "16.63kg",
      },
    },
    Object {
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 10,
            "name": "Quick Attack",
            "type": "Normal",
          },
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
        ],
        "special": Array [
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
          Object {
            "damage": 25,
            "name": "Flame Charge",
            "type": "Fire",
          },
          Object {
            "damage": 40,
            "name": "Body Slam",
            "type": "Normal",
          },
        ],
      },
      "classification": "Fox Pokémon",
      "evolutionRequirements": Object {
        "amount": 50,
        "name": "Vulpi",
      },
      "evolutions": Array [
        Object {
          "id": "038",
          "name": "Ninetales",
        },
      ],
      "fleeRate": 0.1,
      "height": Object {
        "maximum": "0.68m",
        "minimum": "0.53m",
      },
      "id": "037",
      "maxCP": 725,
      "maxHP": 831,
      "name": "Vulpix",
      "resistant": Array [
        "Fire",
        "Grass",
        "Ice",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
      ],
      "weaknesses": Array [
        "Water",
        "Ground",
        "Rock",
      ],
      "weight": Object {
        "maximum": "11.14kg",
        "minimum": "8.66kg",
      },
    },
  ],
  "meta": Object {
    "count": 151,
    "limit": 3,
    "skip": 0,
  },
}
`;

exports[`AppController (e2e) should support requesting a page of pokemons filtered by type and favorites 1`] = `
Object {
  "items": Array [
    Object {
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
          Object {
            "damage": 10,
            "name": "Quick Attack",
            "type": "Normal",
          },
        ],
        "special": Array [
          Object {
            "damage": 40,
            "name": "Body Slam",
            "type": "Normal",
          },
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
          Object {
            "damage": 25,
            "name": "Flame Charge",
            "type": "Fire",
          },
        ],
      },
      "classification": "Fox Pokémon",
      "evolutionRequirements": Object {
        "amount": 50,
        "name": "Vulpi",
      },
      "evolutions": Array [
        Object {
          "id": "038",
          "name": "Ninetales",
        },
      ],
      "fleeRate": 0.1,
      "height": Object {
        "maximum": "0.68m",
        "minimum": "0.53m",
      },
      "id": "037",
      "maxCP": 725,
      "maxHP": 831,
      "name": "Vulpix",
      "resistant": Array [
        "Fire",
        "Grass",
        "Ice",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
      ],
      "weaknesses": Array [
        "Water",
        "Ground",
        "Rock",
      ],
      "weight": Object {
        "maximum": "11.14kg",
        "minimum": "8.66kg",
      },
    },
    Object {
      "Previous evolution(s)": Array [
        Object {
          "id": "037",
          "name": "Vulpix",
        },
      ],
      "attacks": Object {
        "fast": Array [
          Object {
            "damage": 10,
            "name": "Quick Attack",
            "type": "Normal",
          },
          Object {
            "damage": 10,
            "name": "Ember",
            "type": "Fire",
          },
        ],
        "special": Array [
          Object {
            "damage": 80,
            "name": "Heat Wave",
            "type": "Fire",
          },
          Object {
            "damage": 55,
            "name": "Flamethrower",
            "type": "Fire",
          },
          Object {
            "damage": 100,
            "name": "Fire Blast",
            "type": "Fire",
          },
        ],
      },
      "classification": "Fox Pokémon",
      "fleeRate": 0.06,
      "height": Object {
        "maximum": "1.24m",
        "minimum": "0.96m",
      },
      "id": "038",
      "maxCP": 2015,
      "maxHP": 2188,
      "name": "Ninetales",
      "resistant": Array [
        "Fire",
        "Grass",
        "Ice",
        "Bug",
        "Steel",
        "Fairy",
      ],
      "types": Array [
        "Fire",
      ],
      "weaknesses": Array [
        "Water",
        "Ground",
        "Rock",
      ],
      "weight": Object {
        "maximum": "22.39kg",
        "minimum": "17.41kg",
      },
    },
  ],
  "meta": Object {
    "count": 151,
    "limit": 20,
    "skip": 0,
  },
}
`;
